specs/agents/tracing-distributed-tracing.md

```diff
@@ -191,6 +191,63 @@ Implementation note: The `traceparent` and `tracestate` specifications only allo
 This implementation guarantees backwards compatibility with our older agents and compatibility with current OpenTelemetry instrumentations.
 We will eventually remove the `elasticapmtraceparent` entirely as soon as we can safely assume that most users have already upgraded all their agents.
 
+### Baggage
+
+To propagate distributed context, we implement the [W3C Baggage](https://www.w3.org/TR/baggage/) specification.
+
+Agents MUST parse, validate, and attach the `baggage` header according [to the W3C specification](https://www.w3.org/TR/baggage/).
+
+In addition to the W3C specification, agents also SHOULD offer users 2 ways to use the values propagated via baggage:
+- Offer an Baggage API to manually manipulate and read baggage values.
+- Offer baggage related configuration to automatically store baggage values on events.
+
+#### Baggage API
+
+Agents SHOULD offer an API with the following functionalities:
+- Adding a new baggage item
+- Changing an existing baggage item
+- Removing an existing baggage item
+- Reading a specific baggage item
+- Reading all baggage items
+
+In case the given language already offers this API as port of the standard library, the agent MUST integrate with the existing library and it MUST avoid adding a proprietary API. In case there is no such API in the standard library, the agent MUST add this API to the proprietary agent API.
+
+#### Baggage related configurations
+
+The following configurations enable users to automatically store baggage items on a given event without any code change. Baggage items with matching keys are always stored as labels.
```

Comment:
Are they stored as labels or otel-attributes? Right now it is the same, but will this change in the future regarding storing attributes at top-level?

Reply:
Good point! Given the fact we rely on the OTel API, I think storing them as otel-attributes is a better fit. There is however 1 issue: there is no otel-attributes on [errors](https://github.com/elastic/apm-server/blob/main/docs/spec/v2/error.json).

I changed the text and currently it states we should use otel-attributes for 1) transactions and 2) spans, and still use labels for errors (since there is nothing better there...)

I have mixed feeling about this. It's unfortunate that our error data model is different here.

Reply:
Are we sure we want to mix `baggage` and `otel.attributes` ?

E.g the otel documentation explicitly states such lifting has to be done manually:

https://opentelemetry.io/docs/concepts/signals/baggage/#baggage-is-not-the-same-as-span-attributes

Semantically origin matters, I rather see `span.baggage: {}` than `span.otel.attributes: {}` or even `span.labels`.

Reply:
That's a good point, but I don't feel we go against the spec here.

Our configs are the no-code equivalent of such lifting from the linked code-snippet. Baggage only defines how to propagate the data, so there is no `span.baggage` in the stored and serialized spans - if users want to store that data, they need to actively lift it - and as the code snippet suggests, attributes is typically a good fit for that - and that's what we do with our configs.

Reply:
I agree with Greg here. 

Let's take a concrete example: A typical use case for baggage would be something like propagating the `user.id` (let's exclude PII aspects for this example) from let's say a `frontend-service`, so that events, traces along a request can be annotated with the `user.id` and then later one could search for all the interactions of a user. So the `frontend-service` would extract that `user.id` somewhere write it as an attribute and then put it into baggage for propagation downstream. Downstream services will than use that information to annotate the data. 

Now, when using that field, the observability user would not care that `user.id` came from a baggage. They would just want to filter traces, logs, etc. by `user.id`. So, I personally agree that `baggage.user.id` would be misleading in that case. 

> How can users inspect that they might have an actor sending a lot of baggage data if we automatically hoist it to attributes? How do they make the distinction?

That's maybe an argument for storing baggage being rather opt-in, not sure:  https://github.com/elastic/apm/pull/820#discussion_r1274973757.

Reply:
We settled on prefixing baggage in the agents weekly. https://github.com/elastic/apm/pull/820#issuecomment-1671024912